internal/traceroute/tcp.go

```diff
@@ -0,0 +1,221 @@
+	// and it's not because of capabilities/exceeded timeout.
+	case err != nil && !isTracerouteError(err):
+		return wrapError(ctx, err, "failed to read ICMP message")
+
+	// User error: we don't have the necessary capabilities
+	// to open a raw socket for reading ICMP messages.
+	case errors.Is(err, errICMPNotAvailable):
+		return wrapError(ctx, err, "ICMP not available for reading")
+
+	// Timeout error: we didn't receive an ICMP message within
+	// the specified timeout, which is expected when routers
+	// do not respond to our traceroute probes.
+	case errors.Is(err, context.DeadlineExceeded):
+		hop := Hop{
+			Latency: time.Since(start),
+			Addr:    HopAddress{IP: "*"},
+			TTL:     target.hopTTL,
+			Reached: false,
+		}
+		log.DebugContext(ctx, "ICMP read timeout exceeded, no response received")
+		span.AddEvent("ICMP read timeout exceeded", trace.WithAttributes(
+			attribute.Bool("traceroute.target.reached", hop.Reached),
+			attribute.Stringer("traceroute.target.hop", hop),
+			attribute.String("traceroute.target.hop.error", err.Error()),
+		))
+		target.hopChan <- hop
+		return nil
+
+	// Expected ICMP message received: we received an ICMP message
+	// indicating that the TTL has expired, which is the expected behavior
+	// of traceroute.
+	default:
+		hop := Hop{
+			Latency: time.Since(start),
+			Addr:    newHopAddress(packet.remoteAddr),
+			Name:    resolveName(packet.remoteAddr),
+			TTL:     target.hopTTL,
+			Reached: packet.reached,
+		}
+		log.DebugContext(ctx, "Received ICMP message", "port", packet.port, "routerAddr", packet.remoteAddr)
+		span.AddEvent("ICMP message received", trace.WithAttributes(
+			attribute.Bool("traceroute.target.reached", hop.Reached),
+			attribute.Stringer("traceroute.target.hop", hop),
+		))
+		target.hopChan <- hop
+		return nil
+	}
+}
+
+// tcpConn represents a TCP connection with a specific port.
+type tcpConn struct {
+	conn net.Conn
+	port int
+}
+
+// dialTCP dials a TCP connection to the given address with the specified TTL.
+func dialTCP(ctx context.Context, addr net.Addr, ttl int, timeout time.Duration) (tcpConn, error) {
+	port := randomPort()
+
+	// Dialer with control function to set IP_TTL
+	dialer := net.Dialer{
+		LocalAddr: &net.TCPAddr{
+			Port: port,
+		},
+		Timeout: timeout,
+		ControlContext: func(_ context.Context, _, _ string, c syscall.RawConn) error {
+			var opErr error
+			if err := c.Control(func(fd uintptr) {
+				opErr = unix.SetsockoptInt(int(fd), unix.IPPROTO_IP, unix.IP_TTL, ttl) // #nosec G115 // The net package is safe to use
+			}); err != nil {
+				return err
+			}
+			return opErr
+		},
+	}
+
+	conn, err := dialer.DialContext(ctx, "tcp", addr.String())
+	switch {
+	case err == nil:
+		return tcpConn{conn: conn, port: port}, nil
+	case errors.Is(err, unix.EADDRINUSE):
+		// If the address is already in use,
+		// we just retry with a new random port.
+		return dialTCP(ctx, addr, ttl, timeout)
+	default:
+		return tcpConn{conn: conn, port: port}, err
+	}
```

Comment:
Note: I could simplify this to the following:
```suggestion
	conn, err := dialer.DialContext(ctx, "tcp", addr.String())
	if errors.Is(err, unix.EADDRINUSE) {
		// If the address is already in use,
		// we just retry with a new random port.
		return dialTCP(ctx, addr, ttl, timeout)
	}
	return tcpConn{conn: conn, port: port}, err
```
Do you think it's more readable that way or should we keep it like it currently is?